### ISSUE TITLE
[Feature]: Use Blitz.js superjson instead of native JSON.serialize/parse

### ISSUE THREAD

**Issue by sergiodxa** [remix-run/remix#656]
### What is the new or updated feature that you are suggesting?

Use [superjson](https://github.com/blitz-js/superjson) to serialize and parse data returned from loaders/actions.

### Why should this feature be included?

Right now Remix uses JSON.stringify and JSON.parse, this means types like Date (see #634) are going to be converted to strings when serializing it but not converted back to Date on the revive step, and other types like BigInt are not even compatible with JSON.stringify, so if an ORM returns one the user needs to manually convert them to a string before returning the data/response inside the loader.

This superjson library created by the Blitz.js team have differents stringify and parse method that it's capable of serializing more types of data and reviving them back to the original class (see the table here https://github.com/blitz-js/superjson#parse), this way if a loader does this:

```ts
return json({ date: new Date() })
```

Then the useLoaderData will get this

```ts
let { date } = useLoaderData<{ date: Date }>()
let fullYear = date.getFullYear() // this will work
```

And other native objects like BigInt, RegExp, Set, Map and Error will also work as valid return values.

---

Additionally, superjson can be extended to serialize and revive other objects, it may be cool if Remix let you customize that somehow, but I think just the default config would be better than the native JSON methods.

**Comment by mikestopcontinues** [comment]
It's [`8.7kB` gzipped](https://bundlephobia.com/package/superjson@1.8.0). I don't think it's worth including. It goes against the "web fundamentals" approach of Remix. Better to rally TC39 to update the JSON spec instead.

**Comment by sergiodxa** [comment]
@mikestopcontinues it's tree-shakeable, so that's the final size, since React only needs one method it should be smaller, but would need to be measured correctly

**Comment by joaopaulobdac** [comment]
Could this be supported by having serialize/parse hooks in the client/server entries?

**Comment by sergiodxa** [comment]
> Could this be supported by having serialize/parse hooks in the client/server entries?

I don't think so, you can't change how Remix parse the data from the first request and for requests it does when navigating. This is something that Remix needs to do internally, or somehow expose hooks to do that which is not possible right now.

**Comment by joaopaulobdac** [comment]
It would be similar to how tRPC does it: https://trpc.io/docs/data-transformers

**Comment by zachtylr21** [comment]
Does somebody need to give the final word on this before someone can make a PR?

**Comment by sergiodxa** [comment]
@zachtylr21 I already made a PR #662

**Comment by zachtylr21** [comment]
@sergiodxa in the future, is there a way to know which issues are already being worked on? I wonder if we should use the "assignees" section so people don't accidentally start working on the same thing. Or is it just assumed that if you create the issue, you create the PR as well?

**Comment by sergiodxa** [comment]
@zachtylr21 honestly, I don't know, I also don't know why GitHub didn't related the issue with the PR (I'm sure `closes #` and `fix #` worked for that), I think assignees makes sense but without someone on the team doing it no one can do it.

**Comment by zachtylr21** [comment]
@sergiodxa Cool, maybe I'll just assume that if someone else created the issue, they're working on it, or I'll just ask them. Btw, I just looked at your PR, I think it's supposed to be "fixes" not "fix".

**Comment by sergiodxa** [comment]
@zachtylr21 just tried with fixes, it doesn't link them, I'm not sure what's happening with GitHub

**Comment by zachtylr21** [comment]
@sergiodxa weird 🤷

**Comment by MichaelDeBoey** [comment]
> I'm not sure what's happening with GitHub

@sergiodxa I think that's because you're not creating a PR against the default branch (which is `main`)

**Comment by sergiodxa** [comment]
> > I'm not sure what's happening with GitHub
> 
> @sergiodxa I think that's because you're not creating a PR against the default branch (which is `main`)

Ahhh that could be the reason, I think the team should set the default branch as dev so people open PRs automatically against that branch instead of main

**Comment by arthurgeek** [comment]
What is the proposed way to deal with this issue?

Right now, this is what I'm doing, but I'm sure there are better ways (I'm new to TypeScript) :

```typescript
export type TransactionJson = Omit<Transaction, 'createdAt'> & {
  createdAt: string
}

export type LoaderResponse = {
  transactions: TransactionJson[]
}

export const loader: LoaderFunction = async ({ params }) => {
  invariant(params.accountId, 'expected params.accountId')

  const transactions = await getTransactions(params.accountId)

  const jsonTransactions = transactions.map((transaction) => {
    return {
      ...transaction,
      createdAt: transaction.createdAt.toISOString(),
    }
  })

  return json({ transactions: jsonTransactions })
}

export default function Transactions() {
  const { transactions } = useLoaderData<LoaderResponse>()

  [...]
}
```

Good thing I need an string representation of that Date inside `Transactions`, otherwise I would need to map that again to parse the `string` as a `Date`.

**Comment by arthurgeek** [comment]
What is the proposed way to deal with this issue?

Right now, this is what I'm doing, but I'm sure there are better ways (I'm new to TypeScript) :

```typescript
export type TransactionJson = Omit<Transaction, 'createdAt'> & {
  createdAt: string
}

export type LoaderResponse = {
  transactions: TransactionJson[]
}

export const loader: LoaderFunction = async ({ params }) => {
  invariant(params.accountId, 'expected params.accountId')

  const transactions = await getTransactions(params.accountId)

  return json({ transactions })
}

export default function Transactions() {
  const { transactions } = useLoaderData<LoaderResponse>()

  [...]
}
```

Good thing I need an string representation of that Date inside `Transactions`, otherwise I would need to map that `Transaction[]` to parse the `string` as a `Date`.

**Comment by mikestopcontinues** [comment]
@arthurgeek Separate from the `superjson` idea, I've always found it better to store timestamps and convert them to `Date` only when necessary. This was partly for JSON, but mostly just because they cause less headaches. That said, nowadays `date-fns` has made it so I almost never find myself reaching for `Date`.

**Comment by mwarnerdotme** [comment]
As someone who uses Maps and Sets in their (backend) code, I would appreciate support for serializing them.

**Comment by donferi** [comment]
Would love this, currently all my prisma date types are incorrect and it's incredibly frustrating.
